docs/weave/tutorial/11-cron.md

```diff
@@ -0,0 +1,272 @@
+Instead, the signature check step bypassed and conditions must be explicitly present in the context when processing given scheduled message.
+
+This could be a security hole if developer implements the CRON feature poorly. Happily only code can schedule a CRON task and decide on conditions, so this should never be an issue.
+
+### Ticker
+
+[Ticker interface](https://github.com/iov-one/weave/blob/master/cron.go#L11-L27) defines the API of the executor of the scheduled jobs.
+
+```go
+// Ticker is an interface used to call background tasks scheduled for
+// execution.
+type Ticker interface {
+    // Tick is a method called at the beginning of the block. It should be
+    // used to execute any scheduled tasks.
+    //
+    // Because beginning of the block does not allow for an error response
+    // this method does not return one as well. It is the implementation
+    // responsibility to handle all error situations.
+    // In case of an error that is an instance specific (ie database
+    // issues) it might be neccessary for the method to terminate (ie
+    // panic). An instance specific issue means that all other nodes most
+    // likely succeeded processing the task and have different state than
+    // this instance. This means that this node is out of sync with the
+    // rest of the network and cannot continue operating as its state is
+    // invalid.
+    Tick(ctx Context, store CacheableKVStore) TickResult
+}
+```
+
+Default Ticker implementation processes messages in order of execution time, starting with the oldest, in short words FIFO. But if you require a LIFO task queue you can implement it easily.
+
+## Tutorial
+
+### Message
+
+In [CreateArticleMsg](https://github.com/iov-one/blog-tutorial/blob/master/x/blog/codec.proto#L92-L104) we defined a field named as `delete_at` that defines the deletion time of the article that will initated by CRON. During the creation of CreateArticleHandler scheduler is defined as a part of the struct:
+
+### Scheduler Handler
+
+```go
+type CreateArticleHandler struct {
+    auth      x.Authenticator
+    ab        *ArticleBucket
+    bb        *BlogBucket
+    dtb       *DeleteArticleTaskBucket
+    scheduler weave.Scheduler
+}
+```
+
+_weave.Scheduler_ will be initiated and passed to the handler on the application layer.
+
+[The code](https://github.com/iov-one/blog-tutorial/blob/master/x/blog/handler.go#L358-L363) that schedules a job to be processed in future is below:
+
+```go
+// schedule delete task
+	if msg.DeleteAt != 0 {
+		deleteArticleMsg := &DeleteArticleMsg{
+			Metadata:  msg.Metadata,
```

Comment:
Hmm can you explain more?

Reply:
`Metadata` is currently only one field structure, containing `schema` field only. Schema field determines which schema version given entity is in.

https://github.com/iov-one/weave/blob/530b1c06180512361d104373f6c8572e1462d3bd/codec.proto#L12-L14

There are two reasons why copying a metadata must never be done:
- two entities, even if representing an entity with the same ID, can be in different schema state. Their metadata will be different.
- metadata is a structure so that it can be extended. There was a proposal [to add modification time](https://github.com/iov-one/weave/issues/661). It is very likely that more attributes will be added. The new field will most likely have a value unique to each instance.

Reply:
Coming back here to close this task https://github.com/iov-one/docs/issues/168.
How do we determine the schema info for the metadata? For example, the current version of `bnsd` does not contain any model with schema other than 1. Let's say if I were to add a new migration then do I have bump every metadata schema value across the module?

Reply:
Schema version is a mechanism to know what format an entity is. 
We version by the extension, because it is easier than by entity type.

All entities start with schema version 1. Let's say you have a `myblog` extension with an entity `BlogEntry`:

```protobuf
message BlogEntry {
    weave.Metadata metadata = 1;
    string content = 2;
}

message Comment {
  // ...
}
```

After using it for some time, you decide that each blog entry needs a title. Before title was included in the content, now it should be a separate field. What you do is to register a migration, telling how to transform each entity (model and message) of `myblog` extension from version 1 to 2. Migration is always by one version.


```protobuf
message BlogEntry {
    weave.Metadata metadata = 1;
    string content = 2; // It will not be used in newer schemas, but must never be removed.
    string title = 3;
    string body = 4;
}
```

```go
func init() {
            // init
            migration.MustRegister(1, &BlogEntry{}, migration.NoModification)
            migration.MustRegister(1, &Comment{}, migration.NoModification) 
            migration.MustRegister(1, &CreateBlogEntry{}, migration.NoModification)   
            migration.MustRegister(1, &CreateComment{}, migration.NoModification)   

            // second version
            migration.MustRegister(2, &BlogEntry{}, func(db weave.ReadOnlyKVStore, entry migration.Migratable) error {
                e = entry.(*BlogEntry)
                chunks := strings.SplitN(e.Content, '\n', 2) // naive
                e.Title, e.Body = chunks[0], chunks[1]
                return nil
            })
            migration.MustRegister(2, &Comment{}, migration.NoModification) 
            migration.MustRegister(2, &CreateBlogEntry{}, migration.NoModification)   
            migration.MustRegister(2, &CreateComment{}, migration.NoModification)   
}
```
Once deployed, send a message to bump the version of schema used by the `myblog` application. Code forever must support all cases of old and new schema.

Once the migration is executed, stored data is not migrated. It is migrated on the fly when fetched from the database using a bucket that supports migrations or when processing a message using a router that supports migration (all our extensions do both). When stored, it will be stored with the current active version (in this case 2).

Does this help?